common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -0,0 +1,171 @@
+      boolean isBinaryCollation) {
+      this.collationName = collationName;
+      this.collator = collator;
+      this.comparator = comparator;
+      this.version = version;
+      this.hashFunction = hashFunction;
+      this.isBinaryCollation = isBinaryCollation;
+
+      if (isBinaryCollation) {
+        this.equalsFunction = UTF8String::binaryEquals;
+      } else {
+        this.equalsFunction = (s1, s2) -> this.comparator.compare(s1, s2) == 0;
+      }
+    }
+
+    /**
+     * Constructor with comparators that are inherited from the given collator.
+     */
+    public CollationInfo(
+      String collationName, Collator collator, String version, boolean isBinaryCollation) {
+      this(
+        collationName,
+        collator,
+        (s1, s2) -> collator.compare(s1.toString(), s2.toString()),
+        version,
+        s -> collator.getCollationKey(s.toString()).hashCode(),
+        isBinaryCollation);
+    }
+  }
+
+  private final CollationInfo[] collatorTable;
+  private final Hashtable<String, Integer> collationNameToIdMap = new Hashtable<>();
+
+  private CollationFactory() {
+    collatorTable = new CollationInfo[4];
+
+    // Binary comparison. This is the default collation.
+    // No custom comparators will be used for this collation.
+    // Instead, we rely on byte for byte comparison.
+    collatorTable[0] = new CollationInfo(
+      "UCS_BASIC",
+      null,
+      UTF8String::binaryCompare,
+      "1.0",
+      UTF8String::binaryHash,
+      true);
+
+    // Case-insensitive UTF8 binary collation.
+    // TODO: Do in place comparisons instead of creating new strings.
+    collatorTable[1] = new CollationInfo(
+      "UCS_BASIC_LCASE",
+      null,
+      (s1, s2) -> s1.toLowerCase().binaryCompare(s2.toLowerCase()), "1.0",
+      (s) -> s.toLowerCase().binaryHash(),
+      false);
+
+    // UNICODE case sensitive comparison (ROOT locale, in ICU).
+    collatorTable[2] = new CollationInfo(
+      "UNICODE", Collator.getInstance(ULocale.ROOT), "153.120.0.0", true);
+    collatorTable[2].collator.setStrength(Collator.TERTIARY);
+
+
+    // UNICODE case-insensitive comparison (ROOT locale, in ICU + Secondary strength).
+    collatorTable[3] = new CollationInfo(
+            "UNICODE_CI", Collator.getInstance(ULocale.ROOT), "153.120.0.0", false);
+    collatorTable[3].collator.setStrength(Collator.SECONDARY);
+
+    for (int i = 0; i < collatorTable.length; i++) {
+      this.collationNameToIdMap.put(collatorTable[i].collationName, i);
+    }
+  }
+
+  private static final CollationFactory instance = new CollationFactory();
+
+  /**
+   * Returns the collation id for the given collation name.
+   */
+  public int collationNameToId(String collationName) {
+    String normalizedName = collationName.toUpperCase();
+    if (collationNameToIdMap.containsKey(normalizedName)) {
+      return collationNameToIdMap.get(normalizedName);
+    } else {
+      throw SparkIllegalArgumentException.apply(
```

Comment:
It's fine to throw `SparkException`. We have error class now and we don't need different exception classes for different error categories.

Reply:
done.